File: number_of_provinces.cpp
class Solution {
public:
    int findCircleNum(vector<vector<int>>& isConnected) {
        std::unordered_set<int> visited;
        int provinces = 0;

        for (int i = 0; i < isConnected.size(); i++) {
            if (visited.find(i) == visited.end()) {
                dfs(i, isConnected, visited);
                provinces++;
            }
        }

        return provinces;        
    }

private:
    void dfs(int city, std::vector<std::vector<int>>& isConnected, std::unordered_set<int>& visited) {
        visited.insert(city);
        for (int cur = 0; cur < isConnected[city].size(); cur++) {
            int connected = isConnected[city][cur];
            if (connected == 1 && visited.find(cur) == visited.end()) {
                dfs(cur, isConnected, visited);
            }
        }
    }
};

File: number_of_provinces.js
/**
 * @param {number[][]} isConnected
 * @return {number}
 */
var findCircleNum = function(isConnected) {
    const visited = new Set();
    let provinces = 0;
    
    const dfs = (city) => {
        visited.add(city);
        for (let cur = 0; cur < isConnected[city].length; cur++) {
            const connected = isConnected[city][cur];
            if (connected && !visited.has(cur)) {
                dfs(cur);
            }
        }
    };
    
    for (let i = 0; i < isConnected.length; i++) {
        if (!visited.has(i)) {
            dfs(i);
            provinces += 1;
        }
    }
    
    return provinces;    
};
